tools/reviewer/webapp/src/app/dashboard/diff/diff-discussion/diff-discussion.component.ts

```diff
@@ -1,36 +1,103 @@
 // How it looks: https://i.imgur.com/cc6XITV.jpg
+
+// There's used some methods and interfaces from file-changes
+// TODO: make code DRY and reuse it from one place
 @Component({
   selector: 'diff-discussion',
   templateUrl: './diff-discussion.component.html',
   styleUrls: ['./diff-discussion.component.scss'],
   providers: [DiffService],
 })
-export class DiffDiscussionComponent {
+export class DiffDiscussionComponent implements OnInit, OnChanges {
   displayedColumns = ['discussions'];
-  @Input() threads: Thread[];
+  allThreads: Thread[];
+
+  @Input() diff: Diff;
   @Input() diffId: number;
 
   constructor(
     public diffService: DiffService,
     public authService: AuthService,
+    private firebaseService: FirebaseService,
+    private notificationService: NotificationService,
   ) { }
 
+  ngOnInit() {
+    this.refreshThreads();
+  }
+
+  ngOnChanges() {
+    this.refreshThreads();
+  }
+
+  refreshThreads(): void {
+    this.allThreads = this.diff
+      .getThreadList()
+      .concat(this.diff.getDiffThreadList());
+    this.sortThreads(this.allThreads);
+  }
+
   getUnresolvedThreads(): number {
-    return this.threads.filter(thread => !thread.getIsDone()).length;
+    return this.allThreads.filter(thread => !thread.getIsDone()).length;
   }
 
   openFile(thread: Thread): void {
+    if (this.isDiffThread(thread)) {
+      // We can't open diff thread
+      return;
+    }
     this.diffService.openFile(thread.getFile(), this.diffId);
   }
 
   getUsername(comment: Comment): string {
     return this.authService.getUsername(comment.getCreatedBy());
   }
+
+  isDiffThread(thread: Thread): boolean {
+    return (thread.getFile() === undefined) ? true : false;
+  }
+
+  // Sort all thread based on timestamp of last comment of the thread
+  sortThreads(threads: Thread[]): void {
+    threads.sort((a, b) => {
+      const aLastIndex: number = a.getCommentList().length - 1;
+      const bLastIndex: number = b.getCommentList().length - 1;
+      const aTimestamp: number = a.getCommentList()[aLastIndex].getTimestamp();
+      const bTimestamp: number = b.getCommentList()[bLastIndex].getTimestamp();
+
+      // Newest on the top
+      return Math.sign(bTimestamp - aTimestamp);
+    });
+  }
+
+  getThreadBackground(thread: Thread): string {
+    return this.isDiffThread(thread) ? 'diff-thread' : 'code-thread';
+  }
+
+  deleteThread(threadIndex: number): void {
+    let diffThreads: Thread[] = this.allThreads.slice();
+    diffThreads.splice(threadIndex, 1);
+    diffThreads = diffThreads.filter(thread => this.isDiffThread(thread));
+    this.diff.setDiffThreadList(diffThreads);
+    this.firebaseService.updateDiff(this.diff).subscribe(() => {
+      this.notificationService.success('Thread is deleted');
+    });
+  }
+
+  // Make thread resolved/unresolved
+  toggleThread(thread: Thread): void {
+    const isDone: boolean = !thread.getIsDone();
+    thread.setIsDone(isDone);
```

Comment:
What's happening here? something like `thread.setIsDone(!thread.getIsDone())` ?
I see that you use the variable `isDone` later too so I guess putting it in a variable is good.

Reply:
I think `isDone` looks more natural for `boolean` variable.
I'll add more comments to make code more clear.